### PR TITLE
release: 0.9.0

### LINKED ISSUE #2991: ☂️ Documentation changes 0.9.0
# Tentative date, 6th September 2022

https://github.com/orgs/rome/projects/3/views/4?query=is%3Aopen+sort%3Aupdated-desc

# Focus of this release
- stabilize configuration file, make sure it's exhaustive but minimal at the same time
- correctness, make sure that parser and linter work as expected, we catch errors early, and we monitor possible performance regressions
- formatter stability
- Node.js API

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Closes #2991 

## Test Plan

Rendered [changelog](https://github.com/rome/tools/blob/release/0.9.0/CHANGELOG.md).
Rendered website.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,65 @@
 # Rome changelog
 
+## 0.9.0
+
+### CLI
+
+- Added `--stdin-file-path` argument, needed when formatting from standard input.
```
Comment (MichaReiser) [inline CHANGELOG.md:7]: As a user, it's unclear to me why this feature or functionality is necessary. Seems rather a technical detail. 

I assume it is for us to know the working directory and infer the file extension? I would then phrase it around that it is now possible to format files passed from stdin with the example you added further down.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,65 @@
...
+- You can now format content from standard input when using the command `rome format`:
+```shell
+echo "function f() { return {} }" | rome format --stdin-file-path example.js
+```
+- Added `--apply-suggested` argument to the `rome check` command, to apply suggested and safe fixes.
+
+### Linter
+
+- Now all rules will emit an error, check the website if you want to [change the severity of a rule](https://rome.tools/#configure-a-rule).
```
Comment (MichaReiser) [inline CHANGELOG.md:16]: Where some rules only emiting warnings before? Or did we add the functionality to change the severity?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,65 @@
...
+- Added `js/noExtraBooleanCast` lint rule.
+- Added `js/noDupeArgs` lint rule.
+- Added `js/noShadowRestrictedNames` lint rule.
+- Added `js/inlineVariable` code action.
```
Comment (MichaReiser) [inline CHANGELOG.md:20]: Let's add links to the rules

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,65 @@
...
+- Fixed issues in the rule `js/noUnusedVariables`. [#3124](https://github.com/rome/tools/pull/3124) [#3060](https://github.com/rome/tools/pull/3060) [#3004](https://github.com/rome/tools/pull/3004)
+- Fixed issues in the rule `js/noNegationElse`. [#2999](https://github.com/rome/tools/issues/2999)
+- Fixed issues in the rule `js/noShoutyConstants`. [#3077](https://github.com/rome/tools/issues/3077)
```
Comment (MichaReiser) [inline CHANGELOG.md:23]: Can we be more specific `Fixed false positives in the rule ...`. Or what exactly was fixed

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,65 @@
...
+
+##### Arrow functions
+
+##### Conditional expressions
+
+##### Parenthesis
+
+##### Arrays
+
+##### Loops
```
Comment (MichaReiser) [inline CHANGELOG.md:41]: Do you want me to go in that much detail? I considered just to just enumerate the syntaxes where we made significant improvement

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,65 @@
...
+
+
+### Configuration
+
+- **BREAKING CHANGE**: removed the second `"rules"` field from a field group.
```
Comment (MichaReiser) [inline CHANGELOG.md:46]: Should this be at the top because it is a breaking change?

File: crates/rome_cli/src/commands/help.rs [hunk 7]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
 const FORMAT_OPTIONS: Markup = markup! {
     "
-    "<Dim>"--indent-style <tabs|space>"</Dim>"             Determine whether the formatter should use tabs or spaces for indentation (default: tabs)
-    "<Dim>"--indent-size <number>"</Dim>"                  If the indentation style is set to spaces, determine how many spaces should be used for indentation (default: 2)
-    "<Dim>"--line-width <number>"</Dim>"                   Determine how many characters the formatter is allowed to print in a single line (default: 80)
-    "<Dim>"--quote-style <single|double>"</Dim>"           Determine whether the formatter should use single or double quotes for strings (default: double)
-    "<Dim>"--quote-properties <as-needed|preserve>"</Dim>" Determine whether the formatter should preserve quotes in object properties (default: as-needed)
-    "<Dim>"--stdin-file-path <string>"</Dim>"              Mandatory argument to use when piping content via standard input, e.g. echo 'let a;' | rome format --stdin-filepath file.js
-"
+    "<Dim>"--write"</Dim>"                                  Write the output of the formatter to the files instead of printing the diff to the console
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:42]: Nit
```suggestion
    "<Dim>"--write"</Dim>"                                  Edit the files in place (beware!) instead of instead of printing the diff to the console
```

File: crates/rome_cli/src/commands/help.rs [hunk 8]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
+    "<Dim>"--skip-errors"</Dim>"                            Skip over files containing syntax errors instead of returning an error
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:43]: What's the meaning of `returning` in the context of a CLI command? 

> Skip over files containing syntax errors instead of emitting an error diagnostic.

File: crates/rome_cli/src/commands/help.rs [hunk 9]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
+    "<Dim>"--indent-style <tabs|space>"</Dim>"              Determine whether the formatter should use tabs or spaces for indentation (default: tabs)
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:44]:
```suggestion
    "<Dim>"--indent-style <tabs|space>"</Dim>"             Change the indention character (default: tabs)
```

File: crates/rome_cli/src/commands/help.rs [hunk 10]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
+    "<Dim>"--indent-size <number>"</Dim>"                   If the indentation style is set to spaces, determine how many spaces should be used for indentation (default: 2)
+    "<Dim>"--line-width <number>"</Dim>"                    Determine how many characters the formatter is allowed to print in a single line (default: 80)
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:46]: Changes the line length where Rome tries to wrap (default: 80)

File: crates/rome_cli/src/commands/help.rs [hunk 11]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
+    "<Dim>"--quote-style <single|double>"</Dim>"            Determine whether the formatter should use single or double quotes for strings (default: double)
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:47]: Changes the quotation character for strings (default: `"`)

File: crates/rome_cli/src/commands/help.rs [hunk 12]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
+    "<Dim>"--quote-properties <as-needed|preserve>"</Dim>"  Determine whether the formatter should preserve quotes in object properties (default: as-needed)
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:48]: Changes when properties in object should be quoted (default: `as-needed`)

File: crates/rome_cli/src/commands/help.rs [hunk 13]
```diff
@@ -39,13 +39,15 @@ const CHECK: Markup = markup! {
...
+    "<Dim>"--stdin-file-path <string>"</Dim>"                Mandatory argument to use when piping content via standard input, e.g. echo 'let a;' | rome format --stdin-file-path file.js
+    "
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:50]: I find this argument confusing as a user. Can I pass any path? What is this path being used for? 

I wonder if we should split the argument:

* --cwd: Argument specifying the working directory of rome (or --project-dir). Rome classic used to have a cwd argument: --cwd <path> Specify a different working directory`
* --stdin-file-extension: The extension of the input file. This still feels rather technical and I'm wondering if it should be a --stdin-language: "js" | "ts" | "TSX" | "jsx" etc.

File: crates/rome_cli/src/commands/help.rs [hunk 14]
```diff
@@ -68,9 +70,7 @@ const FORMAT: Markup = markup! {
 
     INPUTS can be one or more filesystem path, each pointing to a single file or an entire directory to be searched recursively for supported files
 
-"<Emphasis>"OPTIONS:"</Emphasis>"
-    "<Dim>"--write"</Dim>"                       Write the output of the formatter to the files instead of printing the diff to the console
-    "<Dim>"--skip-errors"</Dim>"                 Skip over files containing syntax errors instead of returning an error"
+"<Emphasis>"OPTIONS:"</Emphasis>""
```
Comment (MichaReiser) [inline crates/rome_cli/src/commands/help.rs:73]: Shouldn't these CLI changes be part of its own PR?

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_unused_variables.rs [hunk 15]
```diff
@@ -12,6 +12,17 @@ use rome_rowan::AstNode;
 declare_rule! {
     /// Disallow unused variables.
     ///
+    /// There are two exceptions to this rule:
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/semantic_analyzers/js/no_unused_variables.rs:15]: Shouldn't this be part of a `no_unused_variables` documentation PR?

File: website/src/_includes/docs/formatter.md [hunk 16]
```diff
@@ -198,7 +200,51 @@ function a(b, c) {
...
+object/function/class expression only.
+
+The main reason for diverging is that parenthesizing the whole expression requires that the logic is 
+implemented in any expression that starts with another child expression 
+(tagged template, binary expressions, computed member/assignment, sequence, conditional, etc. ) 
+and it requires traversing the first expression until it reaches an object expression or any expression 
+that doesn't start with another expression. 
+
+This is rather expensive and Rome's approach avoids the expensive traversal except for 
+object/function and class expressions, of which there should be fewer.
```
Comment (MichaReiser) [inline website/src/_includes/docs/formatter.md:226]: This explanation is too focused on technical details. Do we even need the explanation? We're a prettier-like formatter. That doesn't mean we have 100% compatibility and such detailed difference give the impression that all other syntax match perfectly. 

I think it's best to just give the example where Prettier/Rome differs without much of an explanation if we decide to have this section but I'm wondering if this communicates a false sense of how close Rome's formatting is to Prettier's formatting

File: xtask/lintdoc/src/main.rs [hunk 17]
```diff
@@ -309,6 +309,30 @@ fn parse_documentation(
                 writeln!(content)?;
             }
 
+            Event::Start(Tag::Strong) => {
```
Comment (MichaReiser) [inline xtask/lintdoc/src/main.rs:312]: What's this about?

File: CHANGELOG.md (hunk 2 above)
Comment (ematipico) [reply CHANGELOG.md:16]: Before ALL rules were emitting warning. Maybe  I will rephrase it, because only the recommended rules emit errors

Comment (MichaReiser) [reply CHANGELOG.md:16]: > Changed the default severity for recommended rules to "error". You can change the severity in the `rome.json`

File: website/src/_includes/docs/formatter.md (hunk 16 above)
Comment (ematipico) [reply website/src/_includes/docs/formatter.md:226]: I only copied the text in the issue. Feel free to change if you think it's too technical

File: xtask/lintdoc/src/main.rs (hunk 17 above)
Comment (ematipico) [reply xtask/lintdoc/src/main.rs:312]: So support more markdown tags in our generated documentation. Needed in the `noUnusedVariables` rule

File: crates/rome_js_analyze/src/semantic_analyzers/js/no_unused_variables.rs (hunk 15 above)
Comment (ematipico) [reply crates/rome_js_analyze/src/semantic_analyzers/js/no_unused_variables.rs:15]: No, it's documentation. We can't document something that we haven't released.

File: CHANGELOG.md (hunk 5 above)
Comment (ematipico) [reply CHANGELOG.md:41]: I leave it to you. I just created headings based on the PRs closed in this release cycle.

File: crates/rome_cli/src/commands/help.rs (hunk 13 above)
Comment (ematipico) [reply crates/rome_cli/src/commands/help.rs:50]: I can rephrase. 

This argument can't be split into two. The file is needed to the parser and to the diagnostics. We really need a file name. e.g. `foo.js`

File: crates/rome_cli/src/commands/help.rs (hunk 14 above)
Comment (ematipico) [reply crates/rome_cli/src/commands/help.rs:73]: Both. We need to update the website but there was an issue with the indentation in the CLI. It's not a huge change

File: website/src/_includes/docs/formatter.md (hunk 16 above)
Comment (MichaReiser) [reply website/src/_includes/docs/formatter.md:226]: I'm leaning towards deleting the whole section for the reasons I explained above. Are you ok with this?

Comment (ematipico) [reply website/src/_includes/docs/formatter.md:226]: We can if you think it's not relevant to the users.

On the other hand, it's still relevant to us, so maybe we should add a test to cover the difference, like we did for fill array divergence.

File: crates/rome_cli/src/commands/help.rs (hunk 10 above)
Comment (ematipico) [reply crates/rome_cli/src/commands/help.rs:46]: The original text is clearer to me

File: website/src/_includes/docs/formatter.md (hunk 16 above)
Comment (MichaReiser) [reply website/src/_includes/docs/formatter.md:226]: You can look at the prettier compatibility report and the incompatibilities will show up.

File: CHANGELOG.md (hunk 6 above)
Comment (ematipico) [reply CHANGELOG.md:46]: I think it's fine. We have never been strict to "where" the breaking changes are, they are always listed as first thing under their respective area

File: crates/rome_cli/src/commands/help.rs (hunk 10 above)
Comment (MichaReiser) [reply crates/rome_cli/src/commands/help.rs]: It may just be me but I find the word "determine" to be out of context here. It's as if Rome would determine the line width for you, not that the user can specify the line width. @sebmck any ideas how we could improve the help texts?

Comment (ematipico) [reply crates/rome_cli/src/commands/help.rs]: I changed "determine" with "change"

**Comment by MichaReiser** [comment]
I pushed my changes to the formatter changelog

**Comment by leops** [comment]
I pushed documentation changes related to #3143, along with a few advice in the contributing documentation related to building and running the server on Windows

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,23 +1,87 @@
...
+### CLI
+
+- You can now format content from standard input when using the command `rome format`:
+```shell
+echo "function f() { return {} }" | rome format --stdin-file-path example.js
+```
+the argument  `--stdin-file-path` is mandatory when formatting from standard in. The path should represent a
+file name with its extension.
+- Added `--apply-suggested` argument to the `rome check` command, to apply suggested and safe fixes.
+Suggested fixes should be considered **not stable** and applied with care.
```
Comment (MichaReiser) [inline CHANGELOG.md:14]:
```suggestion
Suggested fixes should be considered **unstable** and applied with care. 
```

Why so? It seems a bit confusing that the **recommended** rules are **unstable**. Why are they recommended?

Do we consider all code fixes unstable? If so, I think it would be easier to understand if this is mentioned as its own bullet point. 

* Code fixes are under development and are still unstable. But feel free to open an issue if you run into any problems

File: CHANGELOG.md [hunk 19]
```diff
@@ -1,23 +1,87 @@
...
+- Added [`js/noExtraBooleanCast`](https://rome.tools/docs/lint/rules/noExtraBooleanCast/) lint rule.
+- Added [`js/noDupeArgs`](https://rome.tools/docs/lint/rules/noDupeArgs/) lint rule.
+- Added [`js/noShadowRestrictedNames`](https://rome.tools/docs/lint/rules/noShadowRestrictedNames/) lint rule.
+- Added `js/inlineVariable` code action.
+- Applied various stability fixes to the rule [`js/noUnusedVariables`](https://rome.tools/docs/lint/rules/noUnusedVariables/). [#3124](https://github.com/rome/tools/pull/3124) [#3060](https://github.com/rome/tools/pull/3060) [#3004](https://github.com/rome/tools/pull/3004)
+- Fixed how the suggestion is applied [`js/noNegationElse`](https://rome.tools/docs/lint/rules/noNegationElse/). [#2999](https://github.com/rome/tools/issues/2999)
+- Fixed false positive in the rule [`js/noShoutyConstants`](https://rome.tools/docs/lint/rules/noShoutyConstants/). [#3077](https://github.com/rome/tools/issues/3077)
+- Fixed false positive in the rule [`ts/useShorthandArrayType`](https://rome.tools/docs/lint/rules/useShorthandArrayType/). [#3111](https://github.com/rome/tools/issues/3111)
```
Comment (MichaReiser) [inline CHANGELOG.md:65]:
```suggestion
- Fixed a false positive in the rule [`js/noShoutyConstants`](https://rome.tools/docs/lint/rules/noShoutyConstants/). [#3077](https://github.com/rome/tools/issues/3077)
- Fixed a false positive in the rule [`ts/useShorthandArrayType`](https://rome.tools/docs/lint/rules/useShorthandArrayType/). [#3111](https://github.com/rome/tools/issues/3111)
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -33,9 +97,15 @@ from the LSP.
   ```
 - Fixed some edge cases where the comment suppressions were not working as expected.
 
+### Linter
+
+The linter is now marked as "alpha" and it can be used to lint code from the CLI and
+from the LSP.
```
Comment (MichaReiser) [inline CHANGELOG.md:103]: How about moving the section about the unstable code fixes to here?

File: CHANGELOG.md [hunk 21]
```diff
@@ -33,9 +97,15 @@ from the LSP.
...
+
+
 ### VSCode
 
-- **BREAKING CHANGE**: Removed the majority of settings that were available in the extension, use the 
+- **BREAKING CHANGE**: Removed the majority of settings that were available in the extension, use the
```
Comment (MichaReiser) [inline CHANGELOG.md:108]:
```suggestion
- **BREAKING CHANGE**: Removed the majority of settings that were available in the extension, in favor of the
```

File: website/src/_includes/docs/architecture.md [hunk 22]
```diff
@@ -0,0 +1,8 @@
+## Architecture (in progress)
+
+Rome uses a server-client architecture to run its tasks.
+
+### Daemon
+
+A [daemon](https://en.wikipedia.org/wiki/Daemon_(computing)) is a long-running server 
+that Rome will spawn in the background, and will use it to process a series tasks as it sees fit.
```
Comment (MichaReiser) [inline website/src/_includes/docs/architecture.md:8]:
```suggestion
that Rome spawns in the background and uses to process requests from the editor and CLI. 
```

**Review by MichaReiser — APPROVED** [review]

File: CHANGELOG.md (hunk 20 above)
Comment (ematipico) [reply CHANGELOG.md:103]: The change is part of the previous release. I just moved its headings to have the same order in each release

File: CHANGELOG.md (hunk 18 above)
Comment (leops) [reply CHANGELOG.md:14]: We could use a different wording here as "unstable" may seem related to the current status of the Rome linter, when this is an advice that applies generally to all code fixes: "Safe fixes" are code actions we're absolutely certain will not cause any syntax error or introduce changes in the behavior of the code, but due to the complex semantics of JavaScript most of code actions we emit are "Suggested fixes" that do not guarantee the same strict level of correctness.
Over time we'll certainly improve the existing actions and we may be able to raise the applicability level for some of them, but I think we'll always have actions we're not entirely certain about that the end user will have to either review manually, or explicitly allow with a flag like `--apply-suggested` if they know what they're doing.

Comment (ematipico) [reply CHANGELOG.md:14]: The concept of suggested fixes and safe fixes is not related to the concept of recommendation of the rules. A rule can be recommended but its fix may be suggested. A rule can be not recommended by us but its fix might be safe.  

Some fixes follows under the category of "suggested" due to the nature of JavaScript as dynamic language, and they are not bullet proof.
